File: frontend/src/types/handover.ts
import type { SBARStructure } from "./sbar";

export type InconsistencyType = "VITAL_TREND" | "MED_STATE" | "ALLERGY_COLLISION";
export type InconsistencySeverity = "INFO" | "WARN" | "CRITICAL";

export interface Inconsistency {
  type: InconsistencyType;
  severity: InconsistencySeverity;
  message: string;
  evidence: Record<string, unknown>;
}

export interface CompletenessScore {
  score: number;
  missing_items: string[];
  details: Record<string, boolean>;
}

export interface HandoverCard {
  card_id: string;
  patient_id: string;
  recorded_at: string;
  raw_transcript: string;
  sbar: SBARStructure;
  inconsistencies: Inconsistency[];
  completeness: CompletenessScore;
  signed: boolean;
  signed_at?: string;
  signed_by?: string;
}

export interface TranscribeResponse {
  transcript: string;
  duration_ms: number;
  provider: "gradium" | "whisper-local" | "browser-webspeech";
}

export interface StructureResponse {
  card_id: string;
  sbar: SBARStructure;
  inconsistencies: Inconsistency[];
  completeness: CompletenessScore;
}

export interface ConfirmMissingRequest {
  item: string;
  value: "explicitly_none" | "addressed_separately" | "not_applicable";
}

export interface SignResponse {
  card_id: string;
  signed_at: string;
  audit_hash: string;
}

File: backend/app/models/handover.py
from __future__ import annotations

import uuid
from datetime import datetime, timezone
from typing import Literal

from pydantic import BaseModel, Field

from app.models.sbar import SBARStructure


class Inconsistency(BaseModel):
    type: Literal["VITAL_TREND", "MED_STATE", "ALLERGY_COLLISION"]
    severity: Literal["INFO", "WARN", "CRITICAL"]
    message: str  # German, physician-readable
    evidence: dict[str, str | float | dict[str, str | float]]


class CompletenessScore(BaseModel):
    score: float  # 0.0 – 1.0
    missing_items: list[str]
    details: dict[str, bool] = Field(default_factory=dict)


class HandoverCard(BaseModel):
    card_id: str = Field(default_factory=lambda: str(uuid.uuid4())[:8])
    patient_id: str
    recorded_at: str = Field(
        default_factory=lambda: datetime.now(timezone.utc).isoformat()
    )
    raw_transcript: str
    sbar: SBARStructure
    inconsistencies: list[Inconsistency] = Field(default_factory=list)
    completeness: CompletenessScore
    signed: bool = False
    signed_at: str | None = None
    signed_by: str | None = None


class HandoverSession(BaseModel):
    session_id: str = Field(default_factory=lambda: str(uuid.uuid4())[:8])
    started_by: str = "Dr. Müller"
    started_at: str = Field(
        default_factory=lambda: datetime.now(timezone.utc).isoformat()
    )
    patient_ids: list[str] = Field(default_factory=list)
    card_ids: list[str] = Field(default_factory=list)
